File: tree/tree.py
from collections import deque
from typing import Optional, TypeVar

T = TypeVar("T")

class Node:
    def __init__(self, value: T) -> None:
        self.value = value
        self.left: Optional["Node"] = None
        self.right: Optional["Node"] = None

class BinaryTree:
    def __init__(self, value: T) -> None:
        self._root = Node(value)

    def get_root(self) -> Node:
        return self._root
    
    def insert(self, value: T) -> None:
        self._insert_node(self._root, Node(value))
    
    def delete(self, value: T) -> Optional[Node]:
       if self._root is None:
           return None
    def _delete_node(self, x: Node, value: T):
        if x is None:
            return None
        if x.value < value:
            self._delete_node(x.left, value)
        elif x.value > value:
            self._delete_node(x.right, value)
        else:
            if x.right is None:
                return x.left
            if x.left is None:
                return x.right
            # compute the min
            t = x.right
            stack = deque([t])
            min_value = None
            while stack:
                min_value = stack.pop()
                if min_value.left:
                    stack.append(min_value.left)
                    
                




    def _insert_node(self, root: Node, value: T) -> Node:
        if root is None:
            return Node(value)
        q = deque([root])  
        while q:
            curr: Node = q.popleft()            
            if curr.left is None:
                curr.left = Node(value)
                return root
            else:
                q.append(curr.left)

            if curr.right is None:
                curr.right = Node(value)
                return root
            else:
                q.append(curr.right)

        return root  

File: datastructures/tree/tree.ts
type Nullable<T> = T | null;

class Node<T> {
    value: T;
    left: Nullable<Node<T>>;
    right: Nullable<Node<T>>;

    constructor(value: T) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

export class BinaryTree<T> {
    private _root: Node<T>;

    constructor(value: T) {
        this._root = new Node(value);
    }

    getRoot(): Node<T> {
        return this._root;
    }

    insert(value: T): void {
        this._insertNode(this._root, value);
    }

    delete(value: T): Nullable<Node<T>> {
        if (!this._root) {
            return null;
        }
        this._root = this._deleteNode(this._root, value);
        return this._root;
    }

    private _deleteNode(x: Nullable<Node<T>>, value: T): Nullable<Node<T>> {
        if (x === null) return null;

        if (x.value < value) {
            x.right = this._deleteNode(x.right, value);
        } else if (x.value > value) {
            x.left = this._deleteNode(x.left, value);
        } else {
            if (x.right === null) return x.left;
            if (x.left === null) return x.right;

            // Find the minimum node in the right subtree
            let t = x.right;
            while (t.left !== null) {
                t = t.left;
            }

            // Replace current node's value with that of t
            x.value = t.value;
            // Delete the duplicate from the right subtree
            x.right = this._deleteNode(x.right, t.value);
        }

        return x;
    }

    private _insertNode(root: Nullable<Node<T>>, value: T): Node<T> {
        if (!root) return new Node(value);

        const queue: Node<T>[] = [root];

        while (queue.length > 0) {
            const curr = queue.shift()!;
            if (curr.left === null) {
                curr.left = new Node(value);
                return root;
            } else {
                queue.push(curr.left);
            }

            if (curr.right === null) {
                curr.right = new Node(value);
                return root;
            } else {
                queue.push(curr.right);
            }
        }

        return root;
    }
}
